src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs

```diff
@@ -224,30 +235,75 @@ impl<'tcx> RegionInferenceContext<'tcx> {
 
         // Get a span
         let (category, span) = categorized_path.first().unwrap();
+
+        match category {
+            ConstraintCategory::AssignmentToUpvar =>
```

Comment:
This is interesting. Looking for an assignment to an upvar is not quite taking the approach I expected, but it also makes sense. I guess though that this code won't fire for something like:

```rust
fn foo(x: &mut Vec<&u32>) {
  bar(|y| x.push(y));
}

fn bar(_: impl FnOnce(&u32)) { }
```

Right?

Reply:
Seems ok, though I can think we can address this case via a similar set of errors, if we generalize the conditions somewhat away from *assignment* and more into looking at the kinds of regions involved. (Or maybe this is just another sort of case to specialize)

Reply:
This is similar to the `issue-7573.rs` case that isn't yet handled. I messed around with making another variant for `ConstraintCategory` that would go through to the `report_closure_error` branch.

Reply:
Fixed with most recent commit.